### PR TITLE
Add translator for Flink DataStreams.

### PR DESCRIPTION
**Pull request by robertwb**
This shares what code is reasonable with the DataSet translation paths.

Timestamp information is preserved, but windowing is not due to the differences in how these are expressed and propagated.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: runners/flink/src/main/java/org/apache/beam/runners/flink/adapter/BeamFlinkDataStreamAdapter.java [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+
+  public BeamFlinkDataStreamAdapter() {
+    this(PipelineOptionsFactory.create());
+  }
+
+  public BeamFlinkDataStreamAdapter(PipelineOptions pipelineOptions) {
+    this.pipelineOptions = pipelineOptions;
+  }
+
+  @SuppressWarnings("nullness")
```
Comment (kennknowles) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/adapter/BeamFlinkDataStreamAdapter.java:72]: Let's not do this with this PR, and ideally fix the prior. These at a minimum need a comment justifying it if you think it is unavoidable, or a specific GitHub Issue to fix them (but we shouldn't just introduce bugs or tech debt unless there is a compelling reason)

Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/adapter/BeamFlinkDataStreamAdapter.java:72]: Yes, I'm updating this just as I did the other.

Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/adapter/BeamFlinkDataStreamAdapter.java]: Pushed.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/adapter/BeamFlinkDataStreamAdapter.java [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    @Override
+    public void processElement(StreamRecord<WindowedValue<T>> element) {
+      output.collect(
+          element.replace(
+              element.getValue().getValue(), element.getValue().getTimestamp().getMillis()));
+    }
+  }
+
+  private <K, V> V getNonNull(Map<K, V> map, K key) {
+    return Preconditions.checkNotNull(map.get(Preconditions.checkNotNull(key)));
```
Comment (kennknowles) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/adapter/BeamFlinkDataStreamAdapter.java:300]: You'll want to use Beam's `checkArgumentNotNull` or `checkStateNotNull` with an explicit error message to indicate the source of the problem. This one from Guava will throw an NPE again bringing users knocking on our door about it being an unreasonable crash with no debugging info.

**Review by robertwb — COMMENTED** [review]
PTAL

File: runners/flink/src/main/java/org/apache/beam/runners/flink/adapter/BeamFlinkDataStreamAdapter.java (hunk 2 above)
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/adapter/BeamFlinkDataStreamAdapter.java]: Done.

**Review by kennknowles — APPROVED** [review]

**Pull request merged**